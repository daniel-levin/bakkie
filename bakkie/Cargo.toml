[package]
name = "bakkie"
version = "0.1.0"
edition = "2024"
license = "Unlicense/MIT"
authors = ["Daniel Levin <daniellevin2607@gmail.com>"]
description = "Simple MCP server framework for Rust"
repository = "https://github.com/daniel-levin/bakkie"
homepage = "https://github.com/daniel-levin/bakkie"
readme = "../README.md"

[dependencies]
serde = { version = "1", features = ["derive"] }
anyhow = "1"
bakkie_derive = "0.1"
bytes = "1.10.1"
serde_json = "1.0.140"
thiserror = "2.0.12"
tokio = { version = "1.46.1", features = ["full"] }
tokio-util = { version = "0.7.15", features = ["full"] }
futures-util = "0.3"
bakkie_schema = "0.1"
schemars = "1.0.4"
tracing = "0.1"
futures = "0.3.31"
strum = { version = "0.27", features = ["derive"] }
monostate = "0.1.14"

[dev-dependencies]
anyhow = "1"
insta = { version = "1", features = ["serde", "json"] }
tracing-subscriber = { version = "0.3", features = ["env-filter"] }
tokio-test = "0.4"
