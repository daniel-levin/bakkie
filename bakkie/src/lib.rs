use thiserror::Error;

pub use bakkie_derive::{structured, tool};
use serde_json::{Map, Value};

pub mod framing;
pub mod proto;
pub mod provisions;

pub mod schemars {
    pub use schemars::*;
}

pub mod serde {
    pub use serde::*;
}

use serde::{Deserialize, Serialize};

pub trait Structured: Sized {
    fn as_json_schema() -> serde_json::Value;

    fn from_json(j: Map<String, Value>) -> Result<Self, serde_json::Error>;

    fn as_json_value(&self) -> serde_json::Value;
}

impl<T> Structured for T
where
    T: schemars::JsonSchema + Serialize + for<'de> Deserialize<'de>,
{
    fn as_json_schema() -> serde_json::Value {
        let schema = schemars::schema_for!(T);
        serde_json::to_value(schema).unwrap()
    }

    fn from_json(j: Map<String, Value>) -> Result<Self, serde_json::Error> {
        serde_json::from_value(Value::Object(j))
    }

    fn as_json_value(&self) -> serde_json::Value {
        serde_json::to_value(self).unwrap()
    }
}

#[derive(Debug, Error)]
#[error(transparent)]
pub struct BakkieError {
    err: BakkieErrorInternal,
}

impl<T: Into<BakkieErrorInternal>> From<T> for BakkieError {
    fn from(t: T) -> Self {
        Self { err: t.into() }
    }
}

pub type Result<T, E = BakkieError> = std::result::Result<T, E>;

#[derive(Debug, Error)]
enum BakkieErrorInternal {
    #[error(transparent)]
    CodecError(#[from] framing::CodecError),
}
pub fn stdio() -> crate::framing::StdioTransport {
    tokio::io::join(tokio::io::stdin(), tokio::io::stdout())
}

#[macro_export]
macro_rules! dnp {
    () => {
        use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
        // Setup tracing to write to named pipe
        let pipe_path = "/tmp/digest_trace";
        let pipe_file = std::fs::OpenOptions::new()
            .create(true)
            .write(true)
            .truncate(true)
            .open(pipe_path)?;

        tracing_subscriber::registry()
            .with(
                tracing_subscriber::fmt::layer()
                    .with_writer(pipe_file)
                    .with_ansi(false),
            )
            .with(
                tracing_subscriber::EnvFilter::from_default_env()
                    .add_directive(tracing::Level::TRACE.into()),
            )
            .init();
    };
}
