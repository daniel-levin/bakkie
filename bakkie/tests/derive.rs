#![allow(unused_imports)]
#![allow(unused_variables)]
#![allow(dead_code)]

use bakkie::{
    framing::{Frame, Msg, Request, RequestId, Transport},
    proto::V20250618::McpServer,
    provisions::{
        Provisions,
        tools::{ToolError, ToolFuture, ToolInput, ToolOutput},
    },
};
use futures::{SinkExt, stream::StreamExt};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use tokio::{io::AsyncWriteExt, sync::mpsc};

static INIT: &str = r#"
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "initialize",
  "params": {
    "protocolVersion": "2025-06-18",
    "capabilities": {
      "roots": {
        "listChanged": true
      },
      "sampling": {}
    },
    "clientInfo": {
      "name": "ExampleClient",
      "version": "1.0.0"
    }
  }
}
"#;

static INITIALIZED: &str = r#"
{
  "jsonrpc": "2.0",
  "method": "notifications/initialized"
}
"#;

#[bakkie::tool(name = "count_letters")]
async fn count_letters(needle: char, haystack: String) -> Result<usize, ToolError> {
    Ok(haystack
        .chars()
        .filter(|c| *c == needle)
        .collect::<Vec<_>>()
        .len())
}

#[tokio::test]
async fn test_macro_generates_struct() {
    let params: serde_json::Map<String, serde_json::Value> = serde_json::from_str(
        r#"
        {
        "needle": "a",
        "haystack": "banana"
        }
        "#,
    )
    .unwrap();

    let x = (count_letters().tool_fn)(ToolInput {
        request_id: RequestId::String("1".into()),
        params,
    })
    .await
    .unwrap();

    let f: ToolOutput = x.into_tool_output();

    dbg!(f);
}

#[test]
fn test_macro_generates_struct_old() {
    // Test that the macro generated a struct and function that works
    let args = count_lettersArgs {
        needle: 'a',
        haystack: "banana".to_string(),
    };

    let result = tokio_test::block_on(async { count_letters_impl(args).await }).unwrap();
    assert_eq!(result, 3); // 'a' appears 3 times in "banana"
}

#[bakkie::structured]
#[derive(Default)]
pub struct Location {
    lat: f64,
    lon: f64,
}

#[bakkie::tool(name = "count_letters")]
async fn remember_location(name: String, location: Location) -> Result<usize, ToolError> {
    todo!();
}

#[test]
fn test_macro_generates_struct2() {
    // Test that the macro generated a struct and function that works
    let args = remember_locationArgs {
        name: "".into(),
        location: Location { lat: 0.0, lon: 0.0 },
    };

    let t: ToolFuture = Box::pin(async move {
        match remember_location_impl(args).await {
            Ok(r) => Ok(Box::new(r) as Box<dyn bakkie::provisions::tools::IntoToolOutput>),
            Err(e) => Err(e),
        }
    });
}

#[bakkie::tool]
async fn test_tool_func(param: String) -> Result<String, ToolError> {
    Ok(param)
}

#[tokio::test]
async fn test_naming_convention() {
    // Now test_tool_func() returns a Tool struct
    let tool = test_tool_func();
    assert_eq!(tool.particulars.name, "test_tool_func");

    // The implementation is available as test_tool_func_impl()
    let args = test_tool_funcArgs {
        param: "test".to_string(),
    };
    let result = test_tool_func_impl(args).await.unwrap();
    assert_eq!(result, "test");
}

#[bakkie::tool]
async fn greet(name: String) -> Result<String, ToolError> {
    Ok(format!("Hello, {name}"))
}

#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
async fn call_derived_tool() -> anyhow::Result<()> {
    let (mut client, server) = tokio::io::duplex(64);

    tokio::task::spawn(async move {
        let provisions = Provisions::default();
        provisions.insert_tool("greet", greet()).await;

        let server = McpServer::new_with_provisions(server, provisions);

        server.run().await
    });

    client.write_all(INIT.as_bytes()).await?;

    let mut framed = client.into_framed();

    let server_hello = framed.next().await;
    assert!(server_hello.is_some());

    let _ = framed
        .send(serde_json::from_str(INITIALIZED).unwrap())
        .await;

    let ask_for_tools: Request = serde_json::from_str(
        r#"
    {
        "jsonrpc": "2.0",
        "id": 2,
        "method": "tools/list",
        "params": {
            "cursor": "optional-cursor-value"
        }
    }"#,
    )
    .unwrap();

    let _ = framed
        .send(Frame::Single(Msg::Request(ask_for_tools)))
        .await;

    let _tools_response = framed.next().await;

    Ok(())
}
