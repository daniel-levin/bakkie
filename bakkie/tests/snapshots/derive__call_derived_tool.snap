---
source: bakkie/tests/derive.rs
expression: tool_response
---
{
  "jsonrpc": "2.0",
  "result": {
    "content": [
      {
        "text": "\"Hello, World\"",
        "type": "text"
      }
    ],
    "structuredContent": {
      "result": "Hello, World"
    }
  },
  "id": 3
}
